### PR TITLE
lib: fix confine_array_index() for 32bit modes

### PR DESCRIPTION
**Pull request by etienne-lms**
Fix implementation of confine_array_index() for 32bit Arm and Thumb2
modes as previous implementation does not explicit the return value
and show issues with GCC 8.3 compiler.

Fixes: https://github.com/OP-TEE/optee_os/issues/3799
Fixes: 2b6dd0df52b4 ("confine_array_index.h: add A32 and T32 versions of confine_array_index()")
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (qemu,qemu_v8)

### CODE REVIEW

**Comment by Emantor** [comment]
`Acked-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>`

File: lib/libutils/ext/include/confine_array_index.h [hunk 1]
```diff
@@ -86,7 +86,7 @@ static inline size_t confine_array_index(size_t index, size_t size) {
 #ifdef __arm__
 static inline size_t confine_array_index(size_t index, size_t size)
 {
-	size_t safe_index = 0;
+	size_t safe_index = index;
```
Comment (jenswi-linaro) [inline lib/libutils/ext/include/confine_array_index.h:89]: The name `safe_index` is  misleading, how about renaming it to something neutral like `ret_val` instead?

Comment (etienne-lms) [reply lib/libutils/ext/include/confine_array_index.h:89]: ok

**Comment by etienne-lms** [comment]
comments addressed. tag applied.

**Comment by etienne-lms** [comment]
@jforissier, I may need your help here.
Shippable fails to build with Clang. Trace below:
```bash
_make PLATFORM=vexpress-qemu_armv8a CFG_ARM64_core=y COMPILER=clang
In file included from ta/pkcs11/src/pkcs11_token.c:7:
out/export-ta_arm32/include/confine_array_index.h:102:3: error: deprecated instruction in IT block [-Werror,-Winline-asm]
        "movcs  %0, #0\n"
         ^
<inline asm>:4:1: note: instantiated into assembly here
movcs   r8, #0
^
1 error generated.
mk/compile.mk:157: recipe for target 'out/ta/pkcs11/src/pkcs11_token.o' failed
make: *** [out/ta/pkcs11/src/pkcs11_token.o] Error 1
make: *** Waiting for unfinished jobs....

**Comment by jforissier** [comment]
@etienne-lms that's interesting because GCC (`gas`) outputs similar warnings in some other places in the code [1] but it seems to ignore the `-Werror` flag so the build doesn't fail ;-)

In this case, could we replace `movcs` with a conditional followed by a `mov`?

A more drastic change would be to ensure we don't emit Thumb code at all. To be honest I'm not sure why/how we have such a mix of Thumb and A32 code.

[1] See for instance the poplar build in https://app.shippable.com/github/OP-TEE/optee_os/runs/4990/1/console
```
{standard input}: Assembler messages:
{standard input}:321: IT blocks containing 32-bit Thumb instructions are performance deprecated in ARMv8-A and ARMv8-R
```

**Comment by etienne-lms** [comment]
This inline function is expected to follow build directives, not to impose Thumb2/Arm mode, no?

**Comment by jforissier** [comment]
Yes, all the message is saying is you shouldn't be using `movcs` in Thumb mode for performance reasons.

**Comment by etienne-lms** [comment]
This is were performance hits security as for come platforms, we need the `movcs followed by `csdb`.
Maybe introduce a configuration switch `CFG_CONFINE_ARRAY_INDEX=y|n` enabled only for impacted cortext-a architectures?

**Comment by jforissier** [comment]
> This is were performance hits security as for come platforms, we need the `movcs followed by `csdb`. Maybe introduce a configuration switch `CFG_CONFINE_ARRAY_INDEX=y|n` enabled only for impacted cortext-a architectures?

That would Introduce more complexity and I think we can avoid that. How about make it a normal (not inline) function and build the file with a flag to disable the warning?

**Comment by jenswi-linaro** [comment]
Perhaps we can just turn of the warning with some magic attribute for the function?

**Comment by jforissier** [comment]
Seems doable:
https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas

**Comment by jforissier** [comment]
@etienne-lms this seems to do the trick:
```diff
diff --git a/lib/libutils/ext/include/confine_array_index.h b/lib/libutils/ext/include/confine_array_index.h
index 99363fec..4f8b1e3b 100644
--- a/lib/libutils/ext/include/confine_array_index.h
+++ b/lib/libutils/ext/include/confine_array_index.h
@@ -99,7 +99,15 @@ static inline size_t confine_array_index(size_t index, size_t size)
        ".syntax unified\n"
        "cmp    %0, %1\n"
        "it     cs\n"
+#ifdef __clang__
+#pragma clang diagnostic push
+        /* Avoid 'deprecated instruction in IT block [-Werror,-Winline-asm]' */
+#pragma clang diagnostic ignored "-Winline-asm"
+#endif
        "movcs  %0, #0\n"
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
        ".inst.n 0xf3af\t@ CSDB\n"
        ".inst.n 0x8014\t@ CSDB"
        : "+r" (ret_val) : "r" (size) : "cc");
```

**Comment by jenswi-linaro** [comment]
Looks good

**Comment by etienne-lms** [comment]
let's wait Shippeable.
i'll apply tags afterward.

**Comment by jforissier** [comment]
s/lib:/libutils:/ no?

**Comment by etienne-lms** [comment]
updated.
@jenswi-linaro , @jforissier, I applied your Acked-by tag on the later commit.

**Comment by jforissier** [comment]
> @jforissier, I applied your Acked-by tag on the later commit.

That's fine. One small thing though: would you mind removing the accents in my name? (`s/Jérôme/Jerome/`). I never use accented characters in my tags, because they are sometimes rendered incorrectly. Maybe it is an outdated habit now that most software use UTF-8 etc. but  let's be consistent ;-)

**Comment by etienne-lms** [comment]
updated

**Pull request merged**